### PR TITLE
Refactor context evaluation & ProbeDetails

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do
Delegate context evaluation depending on the probe type. So we introduced an evaluate method in ProbeDefinition and move current code into LogProbe. Other kinds of probe can then evaluate the context differently depending on what would need to be evaluated. Refactor ProbeDetails to be just an interface so ProbeDefinition will implement is instead of creating a custom class containing all information that are required for the context evaluation. Remove ProbeDefinition.MethodLocation to keep only one in bootstrap project (Used by snapshot and CapturedContext).

# Motivation
Preparing Span Decoration Probe instrumentation

# Additional Notes

### CODE REVIEW

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/ConfigurationUpdater.java [hunk 1]
```diff
@@ -242,34 +241,13 @@ public Snapshot.ProbeDetails resolve(String id, Class<?> callingClass) {
...
-    return new Snapshot.ProbeDetails(
-        probe.getProbeId().getId(),
-        probe.getProbeId().getVersion(),
-        location,
-        ProbeDefinition.MethodLocation.convert(probe.getEvaluateAt()),
-        logProbe.isCaptureSnapshot(),
-        logProbe.getProbeCondition(),
-        probe.concatTags(),
-        new LogMessageTemplateSummaryBuilder(logProbe));
+    definition.buildLocation(type, method);
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/ConfigurationUpdater.java:249]: Not sure why this need to be here - keep building a new location every time we resolve the probe.

Can we buildLocation when instrumenting the probe?

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/ProbeDefinition.java [hunk 2]
```diff
@@ -131,6 +133,64 @@ public abstract void instrument(
...
+      return true;
+    }
+    MethodLocation localEvaluateAt = evaluateAt; // MethodLocation.convert(evaluateAt);
+    if (methodLocation == MethodLocation.ENTRY) {
+      return localEvaluateAt == MethodLocation.DEFAULT || localEvaluateAt == MethodLocation.ENTRY;
+    }
+    return localEvaluateAt == methodLocation;
+  }
+
+  protected static boolean executeScript(
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/ProbeDefinition.java:174]: 1. can we remove/move this function?
2. why we pass probeId? should we do this.id?

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/ProbeDefinition.java [hunk 3]
```diff
@@ -131,6 +133,64 @@ public abstract void instrument(
...
+  }
+
+  @Override
+  public void evaluate(
+      Snapshot.CapturedContext context,
+      Snapshot.CapturedContext.Status status,
+      MethodLocation methodLocation) {}
+
+  @Override
+  public SummaryBuilder getSummaryBuilder() {
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/ProbeDefinition.java:148]: why this is public && here? 

it seems something only log-probe need to know about...

**Review by shatzi — APPROVED** [review]
Good progress. Few comments on areas I believe more refactoring could help.

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/sink/DebuggerSinkTest.java [hunk 4]
```diff
@@ -298,37 +334,42 @@ public void addSnapshotWithCorrelationIdsLineProbe() throws URISyntaxException,
...
     snapshot.addLine(line, 25);
     sink.addSnapshot(snapshot);
-    String fixtureContent =
-        getFixtureContent(SINK_FIXTURE_PREFIX + "/snapshotWithCorrelationIdsLineRegex.txt");
-    String regex = fixtureContent.replaceAll("\\n", "");
     sink.flush(sink);
     verify(batchUploader).upload(payloadCaptor.capture(), matches(EXPECTED_SNAPSHOT_TAGS));
     String strPayload = new String(payloadCaptor.getValue(), StandardCharsets.UTF_8);
-    assertTrue(strPayload.matches(regex), strPayload);
+    System.out.println(strPayload);
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/sink/DebuggerSinkTest.java:344]: we keep. help to troubleshoot, specially in CI

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/sink/DebuggerSinkTest.java [hunk 5]
```diff
@@ -146,12 +170,20 @@ public void addNoSnapshots() {
...
     DebuggerSink sink = new DebuggerSink(config, batchUploader);
     sink.addReceived(new ProbeId("1", 2));
-    String fixtureContent = getFixtureContent(SINK_FIXTURE_PREFIX + "/diagnosticsRegex.txt");
-    String regex = fixtureContent.replaceAll("\\n", "");
     sink.flush(sink);
     verify(batchUploader).upload(payloadCaptor.capture(), matches(EXPECTED_SNAPSHOT_TAGS));
     String strPayload = new String(payloadCaptor.getValue(), StandardCharsets.UTF_8);
-    assertTrue(strPayload.matches(regex), strPayload);
+    System.out.println(strPayload);
+    ParameterizedType type = Types.newParameterizedType(List.class, ProbeStatus.class);
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/sink/DebuggerSinkTest.java:177]: almost nothing related to assertOneIntakeRequest. and only used once for this form. not sure worth it

File: dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java [hunk 6]
```diff
@@ -216,128 +207,110 @@ public enum Kind {
...
+
+    SummaryBuilder getSummaryBuilder();
+
+    MethodLocation getEvaluateAt();
+
+    boolean isCaptureSnapshot();
+
+    boolean hasCondition();
+
+    class DummyProbe implements ProbeDetails {
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java:235]: It's used for UNKNOWN probe and InstrumentTheWorld Probe above, so yeah dummy probes :)
Why should we think more about it?

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/LogProbe.java [hunk 7]
```diff
@@ -322,6 +325,47 @@ public void instrument(
...
+  public void evaluate(
+      Snapshot.CapturedContext context,
+      Snapshot.CapturedContext.Status status,
+      MethodLocation methodLocation) {
+    boolean shouldEvaluate = resolveEvaluateAt(methodLocation);
+    if (!shouldEvaluate) {
+      return;
+    }
+    status.setCondition(executeScript(this.probeCondition, context, id));
+    status.setConditionErrors(context.handleEvalErrors(status.getErrors()));
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/LogProbe.java:338]: I can improve the error reporting in a next PR, this is not really the subject of this one.

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/LogProbe.java [hunk 8]
```diff
@@ -322,6 +325,47 @@ public void instrument(
...
+          new Snapshot.EvaluationError(
+              "uncaught exception", throwable.getType() + ": " + throwable.getMessage()));
+    }
+    if (status.getCondition()) {
+      if (summaryBuilder == null) {
+        summaryBuilder = new LogMessageTemplateSummaryBuilder(segments);
+      }
+      switch (methodLocation) {
+        case ENTRY:
+          this.summaryBuilder.addEntry(context);
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/LogProbe.java:351]: yeah I think we even can get rid of the summaryBuilder interface and just keep a `LogTemplateBuilder`. Can we do this a next PR?

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/ConfigurationUpdater.java (hunk 1 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/ConfigurationUpdater.java:249]: right, done!

**Review by ojung — APPROVED** [review]
Just skimmed really quickly because Idan already did a more thorough review (thx).

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/ProbeDefinition.java (hunk 2 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/ProbeDefinition.java:174]: This function was moved several times, now that it is in ProbeDefinition we can remove the static modifier and use directly fields.
I can move it to LogProbe for now, but likely to be reused by other probe in the future...

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/ProbeDefinition.java (hunk 3 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/ProbeDefinition.java:148]: done

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/LogProbe.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (shatzi) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/LogProbe.java:343]: Sounds good.

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/LogProbe.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (shatzi) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/LogProbe.java:356]: Yup.

File: dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java:235]: I can separate the implementation for testing, ITW and UKNOWN but basically duplicating it for each location

**Pull request merged**